src/core/mesh/qgstriangularmesh.cpp

```diff
@@ -193,11 +199,54 @@ void QgsTriangularMesh::update( QgsMesh *nativeMesh, QgsRenderContext *context )
   // CALCULATE SPATIAL INDEX
   mSpatialIndex = QgsMeshSpatialIndex( mTriangularMesh );
 
-  // SET ALL TRIANGLE CCW
+  // SET ALL TRIANGLE CCW AND COMPUTE AVERAGE SIZE
+  finalizeTriangles();
+
+  QgsEventTracing::addEvent( QgsEventTracing::End, "Triangular mesh", "update" );
+
+  return true;
+}
+
+void QgsTriangularMesh::finalizeTriangles()
+{
+  QgsEventTracing::addEvent( QgsEventTracing::Begin, "Triangular mesh", "finalize" );
+  mAverageTriangleSize = 0;
   for ( int i = 0; i < mTriangularMesh.faceCount(); ++i )
   {
-    setTrianglesCounterClockWise( mTriangularMesh.faces[i] );
+    QgsMeshFace &face = mTriangularMesh.faces[i];
+
+    const QgsMeshVertex &v0 = mTriangularMesh.vertex( face[0] );
+    const QgsMeshVertex &v1 = mTriangularMesh.vertex( face[1] );
+    const QgsMeshVertex &v2 = mTriangularMesh.vertex( face[2] );
+
+    QgsRectangle bbox = QgsMeshLayerUtils::triangleBoundingBox( v0, v1, v2 );
+
+    mAverageTriangleSize += std::fmax( bbox.width(), bbox.height() );
+
+    double ux = v1.x() - v0.x();
```

Comment:
could you please add a comment that this block of code ensures consistent orientation of triangles which is necessary for 3D rendering? (it may not be obvious from the code)